tests/learning/test_fault_tolerance.py

```diff
@@ -0,0 +1,191 @@
+    vladimir._interface_signature_object = signature
+
+    vladimir.substantiate_stamp(password=INSECURE_DEVELOPMENT_PASSWORD)
+
+    # With this slightly more sophisticated attack, his metadata does appear valid.
+    vladimir.validate_metadata()
+
+    # However, the actual handshake proves him wrong.
+    with pytest.raises(vladimir.InvalidNode):
+        vladimir.verify_node(blockchain_alice.network_middleware)
+
+
+def test_emit_warning_upon_new_version(ursula_federated_test_config, caplog):
+    lonely_ursula_maker = partial(make_federated_ursulas,
+                                  ursula_config=ursula_federated_test_config,
+                                  quantity=2,
+                                  know_each_other=True)
+    learner = lonely_ursula_maker().pop()
+    teacher, new_node = lonely_ursula_maker()
+
+    new_node.TEACHER_VERSION = learner.LEARNER_VERSION + 1
+
+    learner._current_teacher_node = teacher
+
+    warnings = []
+
+    def warning_trapper(event):
+        if event['log_level'] == LogLevel.warn:
+            warnings.append(event)
+
+    globalLogPublisher.addObserver(warning_trapper)
+
+    learner.learn_from_teacher_node()
+
+    assert len(warnings) == 1
+    warnings[0]['log_format'] == learner.unknown_version_message.format(new_node, new_node.TEACHER_VERSION,
+                                                                        learner.LEARNER_VERSION)
+
+    # Now let's go a little further: make the version totally unrecognizable.
+    crazy_bytes_representation = int(learner.LEARNER_VERSION + 1).to_bytes(2,
+                                                                           byteorder="big") + b"totally unintelligible nonsense"
+
+    response = Response(content=crazy_bytes_representation, status_code=200)
+    learner.network_middleware.get_nodes_via_rest = lambda *args, **kwargs: response
+    learner.learn_from_teacher_node()
+
+    assert len(warnings) == 2
+    warnings[1]['log_format'] == learner.unknown_version_message.format(new_node, new_node.TEACHER_VERSION,
+                                                                        learner.LEARNER_VERSION)
+
+    globalLogPublisher.removeObserver(warning_trapper)
+
+
+def test_node_posts_future_version(federated_ursulas):
+    ursula = list(federated_ursulas)[0]
+    middleware = MockRestMiddleware()
+
+    warnings = []
+
+    def warning_trapper(event):
+        if event['log_level'] == LogLevel.warn:
+            warnings.append(event)
+
+    globalLogPublisher.addObserver(warning_trapper)
+
+    crazy_node = b"invalid-node"
+    middleware.get_nodes_via_rest(ursula.rest_url(),
+                                  certificate_filepath=ursula.certificate_filepath,
+                                  announce_nodes=(crazy_node,))
+    assert len(warnings) == 1
+    future_node = list(federated_ursulas)[1]
+    future_node.TEACHER_VERSION = future_node.TEACHER_VERSION + 10
```

Comment:
Do you think we need some sort of method to increment versions on a running node?

Reply:
Presently, I'm unable to conceive how a node can continue running (and not restart) when changing versions.  At the very least, it will always require re-evaluation of the modules, right?

Reply:
At the moment its not clear to me how to best minimize or automate the updating process, but this seems potentially relevant.

Reply:
But on a *running* node?  Like, literally without stopping the process?  What kind of update can we possibly perform without stopping the process?

I mean, I know that in some really crazy zero-downtime environments people do crazy shit like dynamically re-eval modules and stuff.  But I don't think we're planning to do that.  And if we are, incrementing an integer is the least of our worries.